### DISCUSSION TITLE
How to decode data to generic type?

### DISCUSSION THREAD

**Discussion by kenan-omnifactors** [skiptools/skip#181] [Q&A] [answered]
Hi, 

In iOS I am decoding data to generic type like the below image. But for android it gives the error. When I used #if !SKIP it builds but just works for iOS. How can I decode data to generic type for Android part?

<img width="1146" alt="Ekran Resmi 2024-07-29 15 41 46" src="https://github.com/user-attachments/assets/0529f53f-ab72-4050-a40b-778572636820">

**Comment by marcprux** [comment]
Apologies for the late response on this. This is indeed a shortcoming in Kotlin's generics handling, which Skip doesn't have any good way of overcoming. Your best bet is to include the type itself as a parameter, even though it is redundant, similar to how we do it with [`JSONDecoder`](https://github.com/skiptools/skip-foundation/blob/main/Sources/SkipFoundation/JSONDecoder.swift#L337). So something like:

```swift
// SKIP DECLARE: override fun <TResponse> send(type: KClass<TResponse>): TResponse where TResponse: Any
func send<TResponse>(_ type: TResponse.Type) throws -> TResponse where TResponse: Decodable {
        return try self.impl.unwrap(as: type)
    let (data, _) = try await URLSession. shared data(for: urlRequest)
    return try JSONDecoder().decode(type, from: data)
}
```

**Comment by sturmenta** [comment, marked as answer]
something similar happened to me and I was able to solve it by implementing this piece of code: https://github.com/orgs/skiptools/discussions/11#discussioncomment-11249025

```swift
@inline(__always) func fetchData<T>(_ type: T.Type, endpoint: String) async throws -> T where T: Decodable {
    ...

    do {
        ...
        return try decoder.decode(type, from: data)
    } catch {
        ...
    }
}
```

**Reply by marcprux** [reply]
Yes, this is a better solution. I had forgotten about using `@inline(__always)` to cause it to use generic reification.
